File: houseRobber/houseRobber.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    int rob(vector<int>& nums) {
        if (nums.size() == 0) return 0;
        vector<int> dp(nums.size(), 0);
        for (int i = 0; i < dp.size(); i++) {
            int prev1 = i - 1 >= 0 ? dp[i - 1] : 0;
            int prev2 = i - 2 >= 0 ? dp[i - 2] : 0;
            dp[i] = max (prev2 + nums[i], prev1);
        }
        return dp.back();
    }
};

int main()
{
    int a[6] = {0, 0, 3, 0, 2, 0};
    vector<int> vec(a, a+6);
    Solution s;
    cout << s.rob(vec) << endl;
    return 0;
}

File: 198.houseRobber/houseRobber.java
class Solution {
    public int rob(int[] nums) {
        if (nums.length < 2) return nums[0];
        int[] dp = new int[nums.length];
        dp[0] = nums[0];
        dp[1] = nums[1];
        if (nums.length == 2) return Math.max(dp[0], dp[1]);
        dp[2] = nums[0] + nums[2];
        int res = Math.max(dp[1], dp[2]);
        for (int i = 3; i < dp.length; i++) {
            dp[i] = Math.max(dp[i - 2], dp[i - 3]) + nums[i];
            res = Math.max(res, dp[i]);
        }
        return res;
    }
}
